### PR TITLE
remove declarative and starlark flags

### PR DESCRIPTION
**Pull request by Shell32-Natsu**
#1547 

Rename `kpt fn run` to `kpt fn eval`. Doc updates will be in separate PR.

### CODE REVIEW

File: thirdparty/cmdconfig/commands/fnevalcmd_test.go [hunk 1]
```diff
@@ -293,7 +221,7 @@ apiVersion: v1
 		},
 		{
 			name: "envs",
-			args: []string{"run", "dir", "--env", "FOO=BAR", "-e", "BAR"},
+			args: []string{"eval", "dir", "--env", "FOO=BAR", "-e", "BAR"},
```
Comment (frankfarzan) [inline thirdparty/cmdconfig/commands/fnevalcmd_test.go:224]: Is declarative run being removed in this PR?

Also please updating the command reference as well as part of this issue (can be separate PR).

File: commands/fncmd.go [hunk 2]
```diff
@@ -43,10 +43,10 @@ func GetFnCommand(name string) *cobra.Command {
...
 	}
 
-	run := commands.RunCommand(name)
-	run.Short = fndocs.RunShort
-	run.Long = fndocs.RunShort + "\n" + fndocs.RunLong
-	run.Example = fndocs.RunExamples
+	eval := commands.EvalCommand(name)
+	eval.Short = fndocs.RunShort
+	eval.Long = fndocs.RunShort + "\n" + fndocs.RunLong
+	eval.Example = fndocs.RunExamples
```
Comment (frankfarzan) [inline commands/fncmd.go:49]: Where is the examples and command reference being updated?

See comment below.

File: thirdparty/cmdconfig/commands/fnevalcmd_test.go [hunk 3]
```diff
@@ -229,7 +157,7 @@ apiVersion: v1
 		{
 			name: "custom kind with storage mounts",
 			args: []string{
-				"run", "dir", "--mount", "type=bind,src=/mount/path,dst=/local/",
+				"eval", "dir", "--mount", "type=bind,src=/mount/path,dst=/local/",
```
Comment (frankfarzan) [inline thirdparty/cmdconfig/commands/fnevalcmd_test.go:160]: Where's the behavior change for validating that either a mount or exec are provided?

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: thirdparty/cmdconfig/commands/fnevalcmd_test.go (hunk 1 above)
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd_test.go:224]: > Is declarative run being removed in this PR?
No, this PR only removes the flags. Separate PR for removing declarative.
> Also please updating the command reference as well as part of this issue (can be separate PR).
Will do.

File: thirdparty/cmdconfig/commands/fnevalcmd_test.go (hunk 3 above)
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd_test.go:160]: https://github.com/GoogleContainerTools/kpt/blob/f6fc94d7a3591b9ec8d05c1bec2577be1edd2643/thirdparty/kyaml/runfn/runfn.go#L474

mount is only used with container runtime.

Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd_test.go:160]: I don't see the behavior change for validating in the fn changes doc.

File: thirdparty/cmdconfig/commands/fnevalcmd.go [hunk 4]
```diff
@@ -101,16 +75,16 @@ type RunFnRunner struct {
...
 
-func (r *RunFnRunner) runE(c *cobra.Command, args []string) error {
+func (r *EvalFnRunner) runE(c *cobra.Command, args []string) error {
 	return runner.HandleError(c, r.RunFns.Execute())
 }
 
 // getContainerFunctions parses the commandline flags and arguments into explicit
 // Functions to run.
-func (r *RunFnRunner) getContainerFunctions(c *cobra.Command, dataItems []string) (
+func (r *EvalFnRunner) getContainerFunctions(dataItems []string) (
```
Comment (droot) [inline thirdparty/cmdconfig/commands/fnevalcmd.go:84]: One observation:
The way this code is written is to have a common code for function invocation with declarative/imperative runs. For example, it creates a functionConfig for `image` or `execPath`. Given that we are dropping declarative runs, we can probably simplify this a lot more.

File: thirdparty/cmdconfig/commands/fnevalcmd.go [hunk 5]
```diff
@@ -252,17 +196,8 @@ func toStorageMounts(mounts []string) []runtimeutil.StorageMount {
...
-
-	if !r.EnableExec && r.ExecPath != "" {
-		return errors.Errorf("must specify --enable-exec with --exec-path")
-	}
-
-	if c.ArgsLenAtDash() >= 0 && r.Image == "" &&
-		!(r.EnableStar && (r.StarPath != "" || r.StarURL != "")) && !(r.EnableExec && r.ExecPath != "") {
+func (r *EvalFnRunner) preRunE(c *cobra.Command, args []string) error {
+	if c.ArgsLenAtDash() >= 0 && r.Image == "" && r.ExecPath == "" {
 		return errors.Errorf("must specify --image")
```
Comment (droot) [inline thirdparty/cmdconfig/commands/fnevalcmd.go:201]: add `or --exec-path` to the error message.

File: thirdparty/kyaml/runfn/runfn_test.go [hunk 6]
```diff
@@ -936,6 +936,38 @@ func TestCmd_Execute(t *testing.T) {
...
+
+	// write a test filter to the directory of configuration
+	if !assert.NoError(t, ioutil.WriteFile(
+		filepath.Join(dir, "filter.yaml"), []byte(ValueReplacerYAMLData), 0600)) {
+		return
+	}
+
+	// write a Kptfile to the directory of configuration
+	if !assert.NoError(t, ioutil.WriteFile(
+		filepath.Join(dir, "Kptfile"), []byte("kind: Deployment"), 0600)) {
```
Comment (droot) [inline thirdparty/kyaml/runfn/runfn_test.go:951]: there is a const for `Kptfile` now, we should use that probably.

Comment (droot) [reply thirdparty/kyaml/runfn/runfn_test.go:951]: I found the test to be very unintuitive in this case where `Kptfile` contained a kind Deployment and it got changed to `StatefulSet` : ) [I think this will break eventually when we implement strict validations of Kptfile during `fn eval`]

I would recommend using kptfile as input and output here.

Another thing about test:
it will be good to do a test without `includeMetaResources` where kptfile is not modified.

File: thirdparty/kyaml/runfn/runfn.go [hunk 7]
```diff
@@ -129,8 +133,12 @@ func (r RunFns) getNodesAndFilters() (
 	// save the output dir because we will need it to write back
 	// the same one for reading must be used for writing if deleting Resources
 	var outputPkg *kio.LocalPackageReadWriter
+	matchFilesGlob := kio.MatchAll
+	if r.IncludeMetaResources {
+		matchFilesGlob = append(matchFilesGlob, "Kptfile")
```
Comment (droot) [inline thirdparty/kyaml/runfn/runfn.go:138]: Pl. use constant for `Kptfile`

**Review by droot — CHANGES_REQUESTED** [review]

File: thirdparty/cmdconfig/commands/fnevalcmd.go [hunk 8]
```diff
@@ -1,30 +1,26 @@
...
 
-// GetCatRunner returns a RunFnRunner.
-func GetRunFnRunner(name string) *RunFnRunner {
-	r := &RunFnRunner{}
+// GetEvalFnRunner returns a EvalFnRunner.
+func GetEvalFnRunner(name string) *EvalFnRunner {
+	r := &EvalFnRunner{}
 	c := &cobra.Command{
-		Use:     "run [DIR]",
+		Use:     "eval [DIR]",
```
[the remark this answers is not in the document]
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd.go:23]: DIR is also used to read resources. Do you mean we will use `.` by default?

File: thirdparty/cmdconfig/commands/fnevalcmd.go (hunk 4 above)
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd.go:84]: Yes I agree

File: thirdparty/cmdconfig/commands/fnevalcmd.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (droot) [reply thirdparty/cmdconfig/commands/fnevalcmd.go:23]: my bad, ignore my comment :)

File: thirdparty/cmdconfig/commands/fnevalcmd.go (hunk 4 above)
Comment (droot) [reply thirdparty/cmdconfig/commands/fnevalcmd.go:84]: We don't need to address this immediately, but something to keep in mind. if we end up in situation where lot of changes need to be made to this code, we can then simplify it.

File: thirdparty/cmdconfig/commands/fnevalcmd.go [hunk 9]
```diff
@@ -34,27 +30,11 @@ func GetRunFnRunner(name string) *RunFnRunner {
...
 	r.Command.Flags().BoolVar(
 		&r.DryRun, "dry-run", false, "print results to stdout")
-	r.Command.Flags().BoolVar(
-		&r.GlobalScope, "global-scope", false, "set global scope for functions.")
-	r.Command.Flags().StringSliceVar(
-		&r.FnPaths, "fn-path", []string{},
-		"read functions from these directories instead of the configuration directory.")
 	r.Command.Flags().StringVar(
 		&r.Image, "image", "",
 		"run this image as a function instead of discovering them.")
```
[the remark this answers is not in the document]
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd.go]: done

File: thirdparty/cmdconfig/commands/fnevalcmd.go [hunk 10]
```diff
@@ -74,23 +54,17 @@ func GetRunFnRunner(name string) *RunFnRunner {
...
 
-// RunFnRunner contains the run function
-type RunFnRunner struct {
+// EvalFnRunner contains the run function
+type EvalFnRunner struct {
 	IncludeSubpackages bool
 	Command            *cobra.Command
 	DryRun             bool
-	GlobalScope        bool
 	FnPaths            []string
```
[the remark this answers is not in the document]
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd.go]: done

File: thirdparty/cmdconfig/commands/fnevalcmd.go (hunk 5 above)
Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd.go]: done

File: thirdparty/kyaml/runfn/runfn_test.go (hunk 6 above)
Comment (Shell32-Natsu) [reply thirdparty/kyaml/runfn/runfn_test.go]: Updated the tests

File: thirdparty/kyaml/runfn/runfn_test.go [hunk 11]
```diff
@@ -936,6 +945,69 @@ func TestCmd_Execute(t *testing.T) {
...
+	}
+	if !assert.NoError(t, instance.Execute()) {
+		t.FailNow()
+	}
+	b, err := ioutil.ReadFile(
+		filepath.Join(dir, v1alpha2.KptFileName))
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	assert.NotContains(t, string(b), "foo: baz")
```
Comment (droot) [inline thirdparty/kyaml/runfn/runfn_test.go:1008]: May be we should check that kptfile content are unchanged.

File: thirdparty/cmdconfig/commands/fnevalcmd_test.go (hunk 3 above)
Comment (droot) [reply thirdparty/cmdconfig/commands/fnevalcmd_test.go:160]: What @frankfarzan probably meant is that `mount`, `network` and `as-user` flags only makes sense with `--image` arg not with `--exec-path`.

Comment (Shell32-Natsu) [reply thirdparty/cmdconfig/commands/fnevalcmd_test.go:160]: Added checks for these flags.

**Review by droot — APPROVED** [review]
